### PR TITLE
Functor Support and Local Vector Initialization

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
ChangeList:

- Functor (Both FrontEnd and MidEnd now use FuncExpr)

- Local Vector Initialization (now you can do sth like var vec: vector{Vertex}(int) = 0;)

### CODE REVIEW

File: test/input/functor_multiple_local_vector.gt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+end
+
+func addOne[a: vector{Vertex}(int), b: vector{Vertex}(int)](v: Vertex)
+    b[v] += 1;
+    a[v] += 2*b[v];
+end
+
+func main()
+
+    var simpleArray1 : vector{Vertex}(int) = new vector{Vertex}(int)();
```
Comment (yunmingzhang17) [inline test/input/functor_multiple_local_vector.gt]: Need similar more-local-vector tests for "edgeset apply" and "edgeset applyUpdatePriority" cases. These tests should preferably have actual values in the vectors so that we can check the vectors are passed in correctly.

File: test/input/functor_edgeset_apply.gt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+const simpleArray: vector{Vertex}(int) = 0;
+
+func updateEdge[addition: int](src: Vertex, dest: Vertex)
+    simpleArray[src] += addition;
+
+end
+
+func main()
+    var addition: int = 5;
+    edges.apply(updateEdge[addition]);
```
Comment (yunmingzhang17) [inline test/input/functor_edgeset_apply.gt]: 1. need a similar test for applyUpdatePriority
2. need additional correctness tests that actually pass in a vector instead of an integer because we mostly work with vectors.
3. need additional correctness tests that test pass in both integer and vectors (mixed type and more than one variable expressions).

File: src/midend/merge_reduce_lower.cpp [hunk 3]
```diff
@@ -60,7 +60,11 @@ namespace graphit {
                 return;
             auto target_expr = mir::to<mir::VarExpr>(tensor_read_expr->target);
             merge_reduce_->field_name = target_expr->var.getName();
-            merge_reduce_->scalar_type = mir::to<mir::ScalarType>(mir_context_->getVectorItemType(merge_reduce_->field_name));
+
+            if (mir_context_->getVectorItemType(merge_reduce_->field_name) != nullptr) {
```
Comment (yunmingzhang17) [inline src/midend/merge_reduce_lower.cpp:64]: Add a comment here

File: src/frontend/token.cpp [hunk 4]
```diff
@@ -261,6 +261,31 @@ namespace graphit {
...
+        while(it != tokens.cend()){
+            Token foundToken = *it;
+            if (foundToken.type == tokenType) return true;
+            it++;
+        }
+        return false;
+
+    }
+
+    int TokenStream::findFirstOccurence(Token::Type tokenType) const {
```
Comment (yunmingzhang17) [inline src/frontend/token.cpp:275]: Add a comment here of what this utility function does?

File: src/midend/mir_emitter.cpp [hunk 5]
```diff
@@ -895,6 +924,64 @@ namespace graphit {
         ctx->addElementType(element_type);
     }
 
+    std::string MIREmitter::convertToFunctorArg(mir::Expr::Ptr arg) {
```
Comment (yunmingzhang17) [inline src/midend/mir_emitter.cpp:927]: Add some comments, especially given there are some restrictions on what Expr type this arg can be.

File: src/midend/mir_emitter.cpp [hunk 6]
```diff
@@ -581,15 +588,20 @@ namespace graphit {
...
                 if (apply_expr->change_tracking_field != nullptr)
                     vertexset_apply_expr->tracking_field = fir::to<fir::Identifier>(
                             apply_expr->change_tracking_field)->ident;
                 retExpr = vertexset_apply_expr;
             } else if (apply_expr->type == fir::ApplyExpr::Type::UPDATE_PRIORITY_EXTERN_APPLY) {
                 auto apply_update_priority_expr = std::make_shared<mir::UpdatePriorityExternVertexSetApplyExpr>();
                 apply_update_priority_expr->target = target_expr;
-                apply_update_priority_expr->input_function_name = apply_expr->input_function->ident;
+                apply_update_priority_expr->input_function_name = apply_expr->input_function->name->ident;
+                apply_update_priority_expr->functorArgs = emitFunctorArgs(apply_expr->input_function->args);
```
Comment (yunmingzhang17) [inline src/midend/mir_emitter.cpp:602]: We need a test case for 
edges.applyUpdatePriority(functor[functor_arg])
See https://github.com/GraphIt-DSL/graphit/blob/master/apps/sssp_delta_stepping.gt

**Review by yunmingzhang17 — COMMENTED** [review]
Mostly looking good. A few comments summarized here 

1. Need test cases for applyUpdatePriority with functor (https://github.com/GraphIt-DSL/graphit/blob/master/apps/sssp_delta_stepping.gt).
2. Can you try to run the performance evaluation scripts (on Lanka is fine) to make sure that we are also not hurting the performance of the benchmarks? 
3. Maybe add comments in a few places?

File: include/graphit/frontend/fir.h [hunk 7]
```diff
@@ -1654,6 +1656,22 @@ namespace graphit {
                virtual void copy(FIRNode::Ptr);
         };
 
+        struct FuncExpr : public Expr {
```
Comment (AjayBrahmakshatriya) [inline include/graphit/frontend/fir.h:1659]: Do we need another FIR node for the FuncExpr parameter? Since all these nodes are expressions, can we not use Call and make the name/target as an expression?

Comment (AjayBrahmakshatriya) [inline include/graphit/frontend/fir.h:1659]: Do we need another FIR node for the FuncExpr parameter? Since all these nodes are expressions, can we not use Call and make the name/target as an expression?

File: test/input/experiments/2D_arrays.gt [hunk 8]
```diff
@@ -0,0 +1,17 @@
+element Vertex end
```
Comment (AjayBrahmakshatriya) [inline test/input/experiments/2D_arrays.gt:1]: What are the experiments? Are they part of the real test suite?

File: test/input/experiments/ipnsw.gt [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        node = v;
+        output = true;
+    else
+        output = false;
+    end
+
+end
+
+
+func update_edge<seen: vector{Vertex}(bool), q: vector[100](float), pq: priority_queue, topk_pq: priority_queue>(src: Vertex, dest: Vertex)
```
Comment (AjayBrahmakshatriya) [inline test/input/experiments/ipnsw.gt:55]: The syntax here doesn't match rest of the test cases/parser. Instead of passing the topk_pq can we pass just an index? Since topk_pqs is global.

File: test/input/experiments/ipnsw.gt [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+  seen[dest] = true;
+  var d_neib: float = dist(data[dest],q);
+  var d_worst: float = -1*topk_pq.top().first;
```
Comment (AjayBrahmakshatriya) [inline test/input/experiments/ipnsw.gt:59]: I don't think top() and .first part of the API we expose for PriorityQueues. Nor does it fit with the syntax. 
If these experiments are just things you have tried, they shouldn't be included in the commit.

File: test/input/closeness_centrality_unweighted_functor.gt [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+     checked_local[dst] = checked_local[src] + 1;
+end
+
+func toFilter[checked_local: vector{Vertex}(int)](v : Vertex) -> output : bool
+     output = checked_local[v] == -1;
+end
+
+func main()
+
+    var checked_local : vector{Vertex}(int) = -1;
```
Comment (AjayBrahmakshatriya) [inline test/input/closeness_centrality_unweighted_functor.gt:17]: We have to be careful while moving declarations local. This should work fine here but might break stuff for the GPU backend. Since the arrays need host and device copies, we rely on them being global and create copies for global declarations only. For now this is OK, but we should discuss how to generalize this.

File: include/graphit/midend/mir.h [hunk 12]
```diff
@@ -857,6 +859,7 @@ namespace graphit {
         struct ApplyExpr : public Expr {
             Expr::Ptr target;
             std::string input_function_name = "";
+            std::vector<std::string> functorArgs;
```
Comment (AjayBrahmakshatriya) [inline include/graphit/midend/mir.h:862]: Why are all these strings? They should be `std::vector<mir::Expr>` so that they can be handled by code generator and analyses in a clean way.

File: src/frontend/fir_visitor.cpp [hunk 13]
```diff
@@ -331,6 +331,10 @@ namespace graphit {
             for (auto genericArg : expr->genericArgs) {
                 genericArg->accept(this);
             }
+
```
Comment (AjayBrahmakshatriya) [inline src/frontend/fir_visitor.cpp:334]: The visitors also need to be modified to have calls to accept for rest of the args (from, to ...)

File: include/graphit/midend/mir.h [hunk 14]
```diff
@@ -907,7 +910,9 @@ namespace graphit {
 
         struct EdgeSetApplyExpr : public ApplyExpr {
             std::string from_func = "";
+            std::vector<std::string> fromFuncFunctorArgs;
```
Comment (AjayBrahmakshatriya) [inline include/graphit/midend/mir.h:913]: Instead of adding the args to each ApplyExpr, you should add a FuncExpr to each of these classes. That will encapsulate the args and the name of the function together. I thought FuncExpr was for this reason. I see there is no FuncExpr in MIR. I think if we are going with FuncExpr, we should use it all the way down.

File: src/frontend/parser.cpp [hunk 15]
```diff
@@ -1454,14 +1491,18 @@ namespace graphit {
 // call_expr: ident ['<' endpoints '>'] '(' [expr_params] ')'
```
Comment (AjayBrahmakshatriya) [inline src/frontend/parser.cpp:1491]: Update the comment here to reflect the new syntax.

File: include/graphit/midend/mir_emitter.h [hunk 16]
```diff
@@ -168,6 +168,12 @@ namespace graphit {
 
         void addElementType(mir::ElementType::Ptr);
 
+        std::vector<std::string> emitFunctorArgs(std::vector<fir::Expr::Ptr> functorArgs);
```
Comment (AjayBrahmakshatriya) [inline include/graphit/midend/mir_emitter.h:171]: Let's not use strings here. It will not work if the arguments aren't variables but other expressions.

**Review by AjayBrahmakshatriya — CHANGES_REQUESTED** [review]
I have added comments for class designs and choice of types for members. I think we should fix these first. I can review the changes to the code based on that.

File: test/input/experiments/2D_arrays.gt (hunk 8 above)
Comment (tugsbayasgalan) [reply test/input/experiments/2D_arrays.gt:1]: Those are not part of the test suite. I was just trying to write few applications that use outer loop parallelization and functor. I can remove them or move them to some other folder.

File: test/input/experiments/ipnsw.gt (hunk 9 above)
Comment (tugsbayasgalan) [reply test/input/experiments/ipnsw.gt:55]: will fix that. thanks

File: test/input/experiments/ipnsw.gt (hunk 10 above)
Comment (tugsbayasgalan) [reply test/input/experiments/ipnsw.gt:59]: yep. i will remove them. thanks :)

File: test/input/closeness_centrality_unweighted_functor.gt (hunk 11 above)
Comment (tugsbayasgalan) [reply test/input/closeness_centrality_unweighted_functor.gt:17]: Ah I see. Yep let's do that.

File: include/graphit/midend/mir.h (hunk 12 above)
Comment (tugsbayasgalan) [reply include/graphit/midend/mir.h:862]: I certainly can change them back to mir::Expr. I initially had them mir::Expr for ApplyExpr but apparently edgeset generation requires FROM and TO functor args to be vector of strings because there was a helper function that generates functor call as string -> https://github.com/GraphIt-DSL/graphit/blob/master/src/backend/codegen_cpp.cpp#L1581. If the functor args for FROM and TO were mir::Expr, it would have required more code changes so I was reluctant to change them to mir::Expr and decided to leave them as strings.  
And  I thought it would be more consistent to just have all the mention of functorArgs . Let me know what you think?

File: src/frontend/parser.cpp (hunk 15 above)
Comment (tugsbayasgalan) [reply src/frontend/parser.cpp:1491]: yep thanks :)

File: include/graphit/midend/mir.h (hunk 14 above)
Comment (tugsbayasgalan) [reply include/graphit/midend/mir.h:913]: Yeah that makes sense. I guess we have to rewrite the edgeset generation then? Instead of having a helper functions that return string, we do something else. @yunmingzhang17 what do you think?

File: src/frontend/parser.cpp (hunk 15 above)
Comment (tugsbayasgalan) [reply src/frontend/parser.cpp:1491]: yep thanks

File: src/midend/merge_reduce_lower.cpp (hunk 3 above)
Comment (tugsbayasgalan) [reply src/midend/merge_reduce_lower.cpp:64]: yep thanks :)

File: src/frontend/token.cpp (hunk 4 above)
Comment (tugsbayasgalan) [reply src/frontend/token.cpp:275]: yep

File: src/midend/mir_emitter.cpp (hunk 5 above)
Comment (tugsbayasgalan) [reply src/midend/mir_emitter.cpp:927]: will do.

File: src/midend/mir_emitter.cpp (hunk 6 above)
Comment (tugsbayasgalan) [reply src/midend/mir_emitter.cpp:602]: sounds good! will do

File: include/graphit/frontend/fir.h (hunk 7 above)
Comment (tugsbayasgalan) [reply include/graphit/frontend/fir.h:1659]: I think we certainly can do that. But isn't it more clear if the functor belongs to a different FIR node than call expr? Especially if i decide to push it all the way down to backend

Comment (yunmingzhang17) [reply include/graphit/frontend/fir.h:1659]: This is a bit different from CallExpr. This is more about passing a function as an argument and not calling the function.

File: include/graphit/midend/mir.h (hunk 12 above)
Comment (yunmingzhang17) [reply include/graphit/midend/mir.h:862]: Yeah, it is legacy of a hack I did before (forcing the arguments to be strings). For now, I think this is fine.

File: include/graphit/midend/mir.h (hunk 14 above)
Comment (yunmingzhang17) [reply include/graphit/midend/mir.h:913]: Yeah, I think Ajay is right on this one. We should make a MIR FuncExpr. I don't think this should affect code generation too much? It is pretty similar to what you have now, except the fromFuncFunctorArgs will be encapsulated in a MIR FuncExpr.

Comment (tugsbayasgalan) [reply include/graphit/midend/mir.h:913]: sounds good. will do

File: include/graphit/midend/mir.h (hunk 12 above)
Comment (tugsbayasgalan) [reply include/graphit/midend/mir.h:862]: @yunmingzhang17 do you think we should also make it FuncExpr as well? Since we are doing it for EdgeSetExpr, might as well we do it for ApplyExpr to be consistent

File: include/graphit/midend/mir_emitter.h (hunk 16 above)
Comment (tugsbayasgalan) [reply include/graphit/midend/mir_emitter.h:171]: Ah yeah. For now we are assuming it can only be Var or Literal expressions.

Comment (tugsbayasgalan) [reply include/graphit/midend/mir_emitter.h:171]: but i agree it should be extended to anytime of expressions

File: include/graphit/midend/mir.h (hunk 12 above)
Comment (yunmingzhang17) [reply include/graphit/midend/mir.h:862]: yup, if you make a MIR FuncExpr, then this might be a good idea.

File: include/graphit/midend/mir_emitter.h (hunk 16 above)
Comment (yunmingzhang17) [reply include/graphit/midend/mir_emitter.h:171]: Either way is fine for me. A lot of things might have to be changed to make it work for any expression, which might be non-trivial.

File: src/backend/gen_edge_apply_func_decl.cpp [hunk 17]
```diff
@@ -295,7 +295,7 @@ namespace graphit {
 
 
         // end of from filtering
-        if (apply->to_func != "") {
+        if (apply->to_func) {
```
Comment (yunmingzhang17) [inline src/backend/gen_edge_apply_func_decl.cpp:298]: this is assuming to_func is initialized to nullptr?

File: include/graphit/midend/mir.h [hunk 18]
```diff
@@ -909,10 +926,11 @@ namespace graphit {
...
 
         struct EdgeSetApplyExpr : public ApplyExpr {
-            std::string from_func = "";
-            std::vector<std::string> fromFuncFunctorArgs;
-            std::string to_func = "";
-            std::vector<std::string> toFuncFunctorArgs;
+
+
+            FuncExpr::Ptr from_func;
+            FuncExpr::Ptr to_func;
```
Comment (yunmingzhang17) [inline include/graphit/midend/mir.h:932]: this might not be nullptr by default. We should still initialize this to either nullptr

File: src/backend/codegen_cpp.cpp [hunk 19]
```diff
@@ -431,6 +435,33 @@ namespace graphit {
         }
     }
 
+    void CodeGenCPP::visit(mir::FuncExpr::Ptr funcExpr) {
```
Comment (yunmingzhang17) [inline src/backend/codegen_cpp.cpp:438]: this is pretty neat.

File: src/backend/codegen_cpp.cpp [hunk 20]
```diff
@@ -1710,60 +1722,72 @@ namespace graphit {
...
 
+        apply->target->accept(this);
+        oss << ", ";
 
-        if (apply->from_func != "") {
-            if (mir_context_->isFunction(apply->from_func)) {
-                // the schedule is an input from function
-                // Create functor instance
-                arguments.push_back(genFunctorNameAsArgumentString(apply->from_func, apply->fromFuncFunctorArgs));
+        if (apply->from_func) {
```
Comment (yunmingzhang17) [inline src/backend/codegen_cpp.cpp:1729]: again, this kind of check might not be safe because from_func is not guaranteed to be a nullptr by default. We need to double-check the initialization part?

File: src/backend/codegen_cpp.cpp [hunk 21]
```diff
@@ -1710,60 +1722,72 @@ namespace graphit {
...
+            //TODO look at it again?
+            //apply_expr->push_function_->functorArgs = apply_expr->input_function->functorArgs;
+            if (pushToFunctionExist) {
+                apply_expr->push_function_->accept(this);
+            } else {
+                oss << ", ";
+                apply_expr->push_function_->accept(this);
+            }
 
-        // the edgeset that is being applied over (target)
```
Comment (yunmingzhang17) [inline src/backend/codegen_cpp.cpp:1618]: It's nice to have this part cleaned up.

File: src/backend/codegen_cpp.cpp [hunk 22]
```diff
@@ -1880,29 +1904,6 @@ namespace graphit {
         oss << "]";
     }
 
-    void CodeGenCPP::genStringArgs(std::vector<std::string> functorArgs) {
```
Comment (yunmingzhang17) [inline src/backend/codegen_cpp.cpp]: This function is replaced by the visit function for FuncExpr?

File: include/graphit/midend/mir.h [hunk 23]
```diff
@@ -856,7 +877,7 @@ namespace graphit {
 
         struct ApplyExpr : public Expr {
             Expr::Ptr target;
-            std::string input_function_name = "";
+            FuncExpr::Ptr input_function;
```
Comment (yunmingzhang17) [inline include/graphit/midend/mir.h:880]: again, doubl-check what is the initial value?

File: src/frontend/parser.cpp [hunk 24]
```diff
@@ -1287,7 +1309,7 @@ namespace graphit {
             } else if (tryConsume(Token::Type::WHERE) || tryConsume(Token::Type::FILTER)) {
                 consume(Token::Type::LP);
                 auto where_expr = std::make_shared<fir::WhereExpr>();
-                where_expr->input_func = parseIdent();
+                where_expr->input_func = parseFunctorExpr();
```
Comment (yunmingzhang17) [inline src/frontend/parser.cpp:1312]: this is good refactoring

File: src/midend/mir_emitter.cpp [hunk 25]
```diff
@@ -895,6 +943,32 @@ namespace graphit {
         ctx->addElementType(element_type);
     }
 
+
+    std::vector<mir::Var> MIREmitter::emitVars(std::vector<fir::Argument::Ptr> args) {
```
Comment (yunmingzhang17) [inline src/midend/mir_emitter.cpp:947]: The name of this function is a bit confusing. This is specifically emitting vector of Var from Arguments. Maybe give it a more descriptive name?

Comment (tugsbayasgalan) [reply src/midend/mir_emitter.cpp:947]: got it
